### PR TITLE
rwx volume: unexport and unmount volume before detaching it

### PR DESCRIPTION
**Pull request by derekbit**
Longhorn/longhorn#6829

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -50,7 +50,7 @@ require (
 	github.com/longhorn/go-spdk-helper v0.0.0-20230802035240-e5fe21b6067f
 	github.com/longhorn/longhorn-engine v1.4.0-rc1.0.20230914160943-b42224518443
 	github.com/longhorn/longhorn-instance-manager v1.4.0-rc1.0.20230804185319-08b3613c5311
-	github.com/longhorn/longhorn-share-manager v1.5.1
+	github.com/longhorn/longhorn-share-manager v1.4.0-rc1.0.20231005131108-b629923cd5b4
 	github.com/pkg/errors v0.9.1
```
Comment (innobead) [inline go.mod:54]: Why the version is still v1.4.0-rc1?

File: controller/share_manager_controller.go [hunk 2]
```diff
@@ -430,8 +430,10 @@ func (c *ShareManagerController) detachShareManagerVolume(sm *longhorn.ShareMana
 	log := getLoggerForShareManager(c.logger, sm)
 
 	shareManagerAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeShareManagerController, sm.Name)
-	log.Infof("Removing volume attachment ticket: %v to detach the volume %v", shareManagerAttachmentTicketID, va.Name)
-	delete(va.Spec.AttachmentTickets, shareManagerAttachmentTicketID)
+	if _, ok := va.Spec.AttachmentTickets[shareManagerAttachmentTicketID]; ok {
+		log.Infof("Removing volume attachment ticket: %v to detach the volume %v", shareManagerAttachmentTicketID, va.Name)
```
Comment (innobead) [inline controller/share_manager_controller.go:465]: This is OK, but why there is a chance the `va.Spec.AttachmentTickets` does not have `shareManagerAttachmentTicketID`?

**Review by innobead — CHANGES_REQUESTED** [review]
Some comments need to resolve first.

File: go.mod (hunk 1 above)
Comment (derekbit) [reply go.mod:54]: `v1.4.0-rc1` is the pseudo version of the commit's base version.
```
╭─root@dereksu-dell ~/go/src/github.com/longhorn/longhorn-share-manager ‹master› 
╰─# git log --oneline --graph
* b629923 (HEAD -> master, upstream/master) vendor: add dependencies
* 42ed161 script: support custom TAG
* bb9748b rpc: implement Unmount server and client functions
* 5c7ee3c proto: add Unmount method
* 0a8efe8 build: update bci-base image from 15.4 to 15.5 As part of longhorn/longhorn#6206
* 48398b9 fix(share-manager): remount volume if mount point became ro
* c996c5a fix(encrypt) crypto parameters for RWX volumes
* 0437075 chore: remove dependcybot
* 1729f0f chore: update nfs ganesha version
* a72d496 chore(build): go 1.20
* 54b3892 grpc: return grpc codes accordingly
* 0b94b34 Close shutdown channel after exiting start()
* 97b2de2 vendor: update dependencies
* 1f179f5 Implement filesystem trim functionality
* 18016fa Start gRPC server when starting daemon
* 863b32e proto: add share manager grpc server
* c2176ae Update dependabot.yml
* cf28a8a chore: add CODEOWNERS file
* 91ddf5a build: use bci image
* 1b79ad7 chore: the io/ioutil package is deprecated from Go 1.16
* 0023fa3 nfs-ganesha: update to v4_20230310
* 8fbe6c0 Dockerfile: update nfs-ganesha to v4.2.3
* 8f41285 Bump github.com/urfave/cli from 1.22.10 to 1.22.12
* 911ac12 Bump github.com/sirupsen/logrus from 1.8.1 to 1.9.0
* f25a9cd Bump github.com/urfave/cli from 1.22.2 to 1.22.10
* e3ec44a ci: add mergify for auto merge
* bb2984f Unify log format
* 177cf45 Update nfs-ganesha to v4.2.1
* 201bc45 Update version
* d54e2c7 (tag: v1.4.0-rc1) Longhorn v1.4.0-rc1 release <------------
* cc490df ci: run CI for release and master branches only
* 7a30b7e Update dependabot.yml
* 2fe1c65 Create dependabot.yml
* b3c9670 build: update bci-base image to 15.4
```


> vX.Y.(Z+1)-0.yyyymmddhhmmss-abcdefabcdef is used when the base version is a release version like vX.Y.Z. For example, if the base version is v1.2.3, a pseudo-version might be v1.2.4-0.20191109021931-daa7c04131f5.

ref https://go.dev/ref/mod#tmp_4

File: controller/share_manager_controller.go (hunk 2 above)
Comment (derekbit) [reply controller/share_manager_controller.go:465]: If the share-manager volume is detached, `shareManagerAttachmentTicketID` is deleted.
The periodic check will print the message constantly.

File: go.mod (hunk 1 above)
Comment (innobead) [reply go.mod:54]: This is not right, we can have a general tag for the master. Let's fix this later. 

```shell
❯ git describe --tags
v1.4.0-rc1-29-gb629923

```

**Review by innobead — APPROVED** [review]

**Pull request merged**